texastribune/salesforce-stripe
# Webpackify

#### What's this PR do?
+ Updates Docker image to include Node, NPM and Yarn
+ Starts tracking dependencies via Yarn and `package.json`
+ Adds front-end build tasks via Webpack
+ Adds dev-only routes via the `FLASK_DEBUG` flag
+ Somehow makes this work with Heroku

#### Why are we doing this? How does it help us?
Prep for the merge with donations-app. And the JS in here is a mess -- it needs ES6-ification.

Having dev-only routes will allow us to continue merging into master during the consolidation process without worrying about those changes appearing in production.

#### How should this be manually tested?
1. `make backing`
2. `make interactive` (this will probably take a while)
3. `yarn run dev`
4. Go to `/devdonate` and confirm it loads properly
5. Change the "Hello world" on line 2 of `/static/js/src/containers/Form.vue` to something else. Confirm Webpack rebuilds it automatically and that, after refreshing, the change is reflected.
6. Verify no new JS files have appeared in `git status`.
7. The past ~25 commits of this branch are me getting the Webpack build working on Heroku. But it finally works! If you have any questions about how that works, happy to demonstrate.

#### Have automated tests been added?
No.

#### Has this been tested on mobile?
NA.

#### Are there performance implications?
NA.

#### What are the relevant tickets?
https://3.basecamp.com/3098728/buckets/736178/todos/1010149314

#### How should this change be communicated to end users?
NA.

#### Next steps?
Get Vue.js working.

#### Smells?
I learned that Heroku can't create directories, meaning that `/static/js/prod/` needs to exist in version control for Heroku to dump compiled files there. That's why it has a `.gitkeep` in it. Kind of smelly -- but it beats versioning the compiled files themselves.

#### TODOs:
* [ ] Enable Node.js build pack in Heroku for the production server

#### Has the relevant documentation/wiki been updated?
Doc strings added, README updated.

#### Technical debt note
Same-ish.
